src/core/lib/iomgr/tcp_cfstream.mm

```diff
@@ -0,0 +1,422 @@
+    grpc_slice_buffer_reset_and_unref_internal(tcp->writeSlices);
+    CallWriteCB(tcp, GRPC_ERROR_CREATE_FROM_STATIC_STRING("Stream closed"));
+    tcp->writeEvent.SetShutdown(GRPC_ERROR_NONE);
+    TCP_UNREF(tcp, "write");
+  } else if (status == kCFStreamStatusError || error != GRPC_ERROR_NONE) {
+    grpc_slice_buffer_reset_and_unref_internal(tcp->writeSlices);
+    if (status == kCFStreamStatusError) {
+      CFErrorRef streamError = CFWriteStreamCopyError(tcp->writeStream);
+      CFWriteStreamClose(tcp->writeStream);
+      CallWriteCB(tcp, GRPC_ERROR_CREATE_FROM_CFERROR(streamError, "Stream error"));
+      CFRelease(streamError);
+    } else {
+      CallWriteCB(tcp, GRPC_ERROR_REF(error));
+    }
+    tcp->writeEvent.SetShutdown(GRPC_ERROR_NONE);
+    TCP_UNREF(tcp, "write");
+  } else if (status == kCFStreamStatusClosed) {
+    grpc_slice_buffer_reset_and_unref_internal(tcp->writeSlices);
+    CallWriteCB(tcp, error);
+    TCP_UNREF(tcp, "write");
+  } else {
+    GPR_ASSERT(CFWriteStreamCanAcceptBytes(tcp->writeStream));
+    grpc_slice slice = grpc_slice_buffer_take_first(tcp->writeSlices);
+    size_t slice_len = GRPC_SLICE_LENGTH(slice);
+    CFIndex writeSize =
+        CFWriteStreamWrite(tcp->writeStream, GRPC_SLICE_START_PTR(slice), slice_len);
+    if (writeSize == -1) {
+      grpc_slice_buffer_reset_and_unref_internal(tcp->writeSlices);
+      CFErrorRef streamError = CFWriteStreamCopyError(tcp->writeStream);
+      CallWriteCB(tcp, GRPC_ERROR_CREATE_FROM_CFERROR(streamError, "write failed."));
+      CFRelease(streamError);
+      TCP_UNREF(tcp, "write");
+    } else {
+      if (writeSize < GRPC_SLICE_LENGTH(slice)) {
+        grpc_slice_buffer_undo_take_first(tcp->writeSlices,
+                                          grpc_slice_sub(slice, writeSize, slice_len));
+      }
+      if (tcp->writeSlices->length > 0) {
+        tcp->writeEvent.NotifyOn(&tcp->writeAction);
+      } else {
+        CallWriteCB(tcp, GRPC_ERROR_NONE);
+        TCP_UNREF(tcp, "write");
+      }
+
+      if (grpc_tcp_trace.enabled()) {
+        grpc_slice trace_slice = slice;
+        if (writeSize < GRPC_SLICE_LENGTH(slice)) {
+          trace_slice = grpc_slice_sub(slice, 0, writeSize);
+        }
+        char* dump = grpc_dump_slice(trace_slice, GPR_DUMP_HEX | GPR_DUMP_ASCII);
+        gpr_log(GPR_DEBUG, "WRITE %p (peer=%s): %s", tcp, tcp->peerString, dump);
+        gpr_free(dump);
+        if (writeSize < GRPC_SLICE_LENGTH(slice)) {
+          grpc_slice_unref(trace_slice);
+        }
+      }
+    }
+    grpc_slice_unref(slice);
+  }
+}
+
+static void ReadCallback(CFReadStreamRef stream, CFStreamEventType type, void* clientCallBackInfo) {
+  CFStreamTCP* tcp = static_cast<CFStreamTCP*>(clientCallBackInfo);
+  TCP_REF(tcp, "read callback");
+  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+    gpr_log(GPR_DEBUG, "TCP:%p readCallback (%p, %lu, %p)", tcp, stream, type, clientCallBackInfo);
```

Comment:
Redispatching to concurrent queue can change the event ordering here, which appears to at least affect logging. Should this be a serial queue?

Reply:
I thought about that too but I tend to leave it concurrent. Read path and write path are relatively independent; at this time I do not see down side to execute them at the same time. As to logging, I think the order of callbacks does not matter that much in this scenario. We simply log the time when the callback gets processes by gRPC subsystem.